### ISSUE TITLE
Unnecessary parentheses in LaTEX output

### ISSUE THREAD

**Issue by m93a** [josdejong/mathjs#1298]
There are many functions in `mathjs` that, when rendered  with `toTex()`, put too many parens around the arguments. For example a simple `cross(c, d)` will produce: `\left( c\right)\times\left( d\right)` instead of `c \times d`. This was already mentioned in #307 where @FSMaxB said he isn't going to have time for that, so it was closed. However some two years later, things could be different. And if not, I could look into the code myself and see what I can do…

Steps needed:
 * [ ] remove parens that enclose a single identifier
   * eg. `not(a)` ⇒ `\neg\left( a\right)` × `\neg a`
 * [ ] remove parens that enclose the whole expression
   * eg. `expm1(x)` ⇒ `\left(e^{ x}-1\right)`) × `e^{ x}-1`
   * the whole output doesn't need to be wrapped in parentheses
 * [ ] remove parens that enclose a fraction
   * eg. `-(1/2)` ⇒ `-\left(\frac{1}{2}\right)` × `-\frac{1}{2}`
   * this feature probably shouldn't be recursive, so that users can force parens around a fraction by writing this: `-((a/b))`
 * [ ] investigate the possibility to detect precedence based on semantics of the expression
   * eg. `add(x, add(y, z))` ⇒  `\left( x+\left( y+ z\right)\right)` × `x+y+z`
   * probably quite hard to implement – I'm not sure whether it's worth the time…

@josdejong What do you think about this?

**Comment by FSMaxB** [comment]
It is still true that I have no time to do this myself but feel free to ask questions anytime!

**Comment by FSMaxB** [comment]
Another approach would be to replace some of the LaTeX templates with actual functions that have their own logic in them. But that probably won't scale because it would be a case by case solution.

**Comment by m93a** [comment]
I think that implementing precedence logic for `FunctionNode` is what I meant by “detect precedence based on semantics of the expression” in the last point. This sounds quite hard and I don't think I'll have time for that either.

On the other hand, putting logic to the `toTex` methods of individual functions would be much easier and could easily solve the first three points. You're right that scalability is worse with this approach, but if we defined some kind of helper function to create templates that take care of that, it could be fine…

**Comment by FSMaxB** [comment]
You don't need to define a helper function to create templates. If I remember correctly you can put a function in there directly instead of the template string (similar or equal to the custom LaTeX handler described in the [documentation](http://mathjs.org/docs/expressions/customization.html#custom-latex-handlers)).

**Comment by FSMaxB** [comment]
This is the relevant code:

https://github.com/josdejong/mathjs/blob/47a291aa62d1916e8cf373d2383e477255699348/src/expression/node/FunctionNode.js#L385-L415

If the `.toTex` property of a function is of type `function`, it get's called with the `FunctionNode` and `options` as parameters instead of expanding a template string.

**Comment by josdejong** [comment]
Thanks for your inputs @m93a . Would be nice to improve the LaTeX output for the cases you reported. We should indeed pay close attention to correctly handling precedence like Max points out.

**Comment by thunderkid** [comment]
Another case that I noticed is where parentheses surround the numerator and denominator of a fraction. So
```math2.parse('(1 - y)/(1+ 2*x)').toTex()```
yields
```\frac{\left(1- y\right)}{\left(1+2\cdot x\right)}```
So you get ugly surrounding brackets on both the numerator and denominator.
It seems that whenever `\left` and `\right` occur as the outermost parts of a `\frac` numerator or denominator, they could be killed off. Unless I'm missing some latex special cases?

**Comment by m93a** [comment]
For anyone interested in working on this, here's how Sympy does it: [sympy/printing/precedence.py](https://github.com/sympy/sympy/blob/62000f37b8821573ba00280524ffb4ac4a380875/sympy/printing/precedence.py)

**Comment by josdejong** [comment]
@thunderkid you're right, in that case the parentheses can be removed from LaTeX output.
